smartsim/_core/control/previewrenderer.py

```diff
@@ -0,0 +1,120 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2023, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import typing as t
+
+import jinja2
+
+from ..._core.config import CONFIG
+from ..._core.control import Manifest
+from ...log import get_logger
+
+logger = get_logger(__name__)
+
+if t.TYPE_CHECKING:
+    from smartsim import Experiment
+
+_OutputFormatString = t.Optional[t.Literal["html"]]
+_VerbosityLevelString = t.Literal["info", "debug", "developer"]
+
+
+def render(
+    exp: "Experiment",
+    manifest: t.Optional[Manifest] = None,
+    verbosity_level: _VerbosityLevelString = "info",
+    output_format: _OutputFormatString = None,
+) -> str:
+    """
+    Render the template from the supplied entities.
+    :param experiment: the experiment to be previewed.
+    :type experiment: Experiment
+    :param manifest: the manifest to be previewed.
+    :type manifest: Manifest
+    :param verbosity_level: the verbosity level
+    :type verbosity_level: _VerbosityLevelString
+    :param output_format: the output destination.
+    If no output format is set, the preview will be output to stdout
+    :type output_format: _OutputFormatString
+    """
+
+    verbosity_level = _check_verbosity_level(verbosity_level)
+
+    loader = jinja2.PackageLoader("templates")
+    env = jinja2.Environment(loader=loader, autoescape=True)
+
+    version = f"_{output_format}" if output_format else ""
+    tpl_path = f"preview/base{version}.template"
+
+    if output_format:
```

Comment:
I dislike that we use `output_format` to indicate both the format AND that it will output to a file. 
- Consider setting the default output_format to `text`
- Consider avoiding a side effect (writing a file) when rendering
  - allow _caller_ to decide what to do with it


Consider this example that lets us add CLI arguments to let the user specify format & path. Who cares if they output text to a file or html to console. If that's what they want, _give it to them_.

```py 
class Experiment:
  ...
  def preview(self, args: arparse.Namespace) -> None:
    format = args.format
    
    # the renderer doesn't render AND write, it just renders
    rendered = previewrenderer.render(self, format=format, ...)
    

    if args.output_path:
        # write file doesn't care if it's a preview, only a path & some content
        # so it's less likely to have to add branching vs nesting it in render code
        self.write_file(args.output_path, content=rendered)
```

Reply:
an alternative might be to pass `output_path` to `previewrenderer.render` to avoid having `format` represent both an output format AND place to render

Reply:
That is what the purpose of output_filename is now. If a user passes a filename, the output will be output to that file instead of to stdout. Do you think that is okay?